File: 0801-0900/0900-RLE Iterator/0900-RLE Iterator.py
class RLEIterator:
    
    def __init__(self, A: List[int]):
        self.A = A
        self.start = 0
        self.num = 0

    def next(self, n: int) -> int:
        while self.start < len(self.A):
            if self.num + n <= self.A[self.start]:
                self.num += n
                return self.A[self.start + 1]
            else:
                n -= self.A[self.start] - self.num
                self.num = 0
                self.start += 2
        return -1


# Your RLEIterator object will be instantiated and called as such:
# obj = RLEIterator(A)
# param_1 = obj.next(n)

File: 0801-0900/0900-RLE Iterator/0900-RLE Iterator.cpp
class RLEIterator {
public:
    RLEIterator(vector<int>& A) {
        this->A = A;
        start = num = 0;
    }
    
    int next(int n) {
        while (start < A.size()) {
            if (num + n <= A[start]) {
                num += n;
                return A[start + 1];
            }
            else {
                n -= A[start] - num;
                num = 0;
                start += 2;
            }
        }
        return -1;
    }
    
private:
    vector<int> A;
    int start, num;
};

/**
 * Your RLEIterator object will be instantiated and called as such:
 * RLEIterator* obj = new RLEIterator(A);
 * int param_1 = obj->next(n);
 */
